File: functions/fsharp/HigherOrderFunctions.fs
module HigherOrderFunctions

let printADoubleNumber number = printfn "%i" (number * 2)

let performActionOnElements list action =
    list |> List.iter action

let numberIsEven number = number % 2 = 0
let getEvenNumbersFromAList =
    [0..10]
    |> List.filter numberIsEven

let printNames names =
    performActionOnElements names (printfn "Hello %s")

let doubleNumbersPrinting numbers =
    performActionOnElements numbers
        (fun number -> printfn "%i" (number * 2))

let sumWithFive number = number + 5
let sumWithFiveViaLambda = fun number -> number + 5

let multiplyAndPrintNumber multiplier number =
    printfn "%i" (number * multiplier)

let printListOfNumbers multiplier numbers =
    performActionOnElements numbers
        (fun number -> multiplyAndPrintNumber multiplier number)

File: functions/csharp/HigherOrderFunctions.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace csharp
{
    static class HigherOrderFunctions
    {
        public static void PerformActionOnElements<ElementType>(List<ElementType> list, Action<ElementType> action)
        {
            list.ForEach(action);
        }

        public static void PrintNames(List<string> names)
        {
            PerformActionOnElements(names, Console.WriteLine);
        }

        public static void DoublePrintingNumbers(List<int> numbers)
        {
            PerformActionOnElements(numbers,
                number => Console.WriteLine(number * 2));
        }

        public static Action<int> MultiplyAndPrintNumbers(int multiplier)
        {
            return (number) => Console.WriteLine(number * multiplier);
        }

        public static void PrintNumbers(List<int> numbers, int multiplier)
        {
            Action<int> multiply = MultiplyAndPrintNumbers(multiplier);
            PerformActionOnElements(numbers,
                number => multiply(number));
        }
    }

    static class CheckerEvenNumber
    {
        public static bool NumberIsEven(int number)
        {
            return number % 2 == 0;
        }

        public static void GetEvenNumbersFromAList()
        {
            Func<int, bool> methodToFilterViaLambda =
                number => number % 2 == 0;
            IEnumerable<int> evenNumbers;
            IEnumerable<int> numbers = Enumerable.Range(0, 10);
            evenNumbers = numbers.Where(methodToFilterViaLambda);

            foreach(int evenNumber in evenNumbers)
            {
                Console.WriteLine(evenNumber);
            }
        }
    }
}
